### PR TITLE
Storm 1290:port backtype.storm.local-state-test to java

### PR DESCRIPTION
**Pull request by kamleshbhatt**
Please review.

### CODE REVIEW

File: storm-server/src/test/java/org/apache/storm/LocalStateTest.java [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+            Assert.assertEquals(globalStreamId_a, ls1.get("a"));
+            Assert.assertEquals(null, ls1.get("c"));
+            Assert.assertEquals(globalStreamId_b, ls1.get("b"));
+            Assert.assertTrue(ls2.snapshot().isEmpty());
+
+            ls2.put("b", globalStreamId_a);
+            ls2.put("b", globalStreamId_b);
+            ls2.put("b", globalStreamId_c);
+            ls2.put("b", globalStreamId_d);
+            Assert.assertEquals(globalStreamId_d, ls2.get("b"));
```
Comment (srdo) [inline storm-server/src/test/java/org/apache/storm/LocalStateTest.java]: Nitpick: The `(is (= {"a" gs-a "b" gs-b} (.snapshot (LocalState. dir1))))` check seems to be missing? Looks like it's checking that creating a new LocalState with the same directory will lead to getting the same state.

**Comment by srdo** [comment]
This LGTM

File: storm-server/src/test/java/org/apache/storm/LocalStateTest.java (hunk 1 above)
Comment (kamleshbhatt) [inline storm-server/src/test/java/org/apache/storm/LocalStateTest.java]: This seems to be duplicate of
ls1 (LocalState. dir1)
(is (= {"a" gs-a "b" gs-b} (.snapshot ls1)))
which is already covered so I left it considering duplicate.

Comment (srdo) [inline storm-server/src/test/java/org/apache/storm/LocalStateTest.java]: I think it was testing that instantiating a new LocalState with the same dir as ls1 would result in the same snapshot being output. An alternative test could be `assertEquals(ls1.snapshot(), new LocalState(dir1_tmp.getPath()).snapshot())`.

**Comment by HeartSaVioR** [comment]
@kamleshbhatt Could you squash commits into one? Too many commits exist for about 100 lines of change.

File: storm-server/src/test/java/org/apache/storm/LocalStateTest.java [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+        }
+    }
+}
```
Comment (HeartSaVioR) [inline storm-server/src/test/java/org/apache/storm/LocalStateTest.java]: It would be better if we port also `empty-state` too.

Comment (kamleshbhatt) [inline storm-server/src/test/java/org/apache/storm/LocalStateTest.java]: The tests in empty-state was already covered in testLocalState. Adding this would be a kind of duplicate to the test cases covered in testLocalState.

Comment (kamleshbhatt) [inline storm-server/src/test/java/org/apache/storm/LocalStateTest.java]: The tests in empty-state was already covered in testLocalState. Adding this would be a kind of duplicate to the test cases covered in testLocalState.

Comment (srdo) [inline storm-server/src/test/java/org/apache/storm/LocalStateTest.java]: I thought so too, but it seems like empty-state is checking that adding some extra files to the LocalState dir isn't interfering with what the LocalState returns. I don't think the other test does that?

Comment (kamleshbhatt) [inline storm-server/src/test/java/org/apache/storm/LocalStateTest.java]: Added the empty-state test.

**Comment by srdo** [comment]
+1

**Comment by harshach** [comment]
@kamleshbhatt can you please squash commits into single commit.

**Comment by harshach** [comment]
+1. once the commits gets squashed.

**Comment by kamleshbhatt** [comment]
@harshach I tried multiple times but unable to squash the commits.

**Comment by srdo** [comment]
@kamleshbhatt You can use soft reset to squash the commits if interactive rebase is too much of a hassle. It might be in this case, because you have a merge commit in the history. In the future consider using `git pull --rebase` when pulling master into your branch, this can make squashing easier later.

Start by copying this branch to a backup `git branch STORM-1290-backup STORM-1290` just to be safe.

Then switch to STORM-1290 `git checkout STORM-1290` and soft reset back to the commit on master you branched from. Since your first commit to this branch is called da1c92c1f1a3c659532ce03d11a3ea376aa16e6c, you want the parent of that commit (in git syntax this is da1c92c1f1a3c659532ce03d11a3ea376aa16e6c~). So do `git reset --soft da1c92c1f1a3c659532ce03d11a3ea376aa16e6c~`. This resets git back to the commit you branched from, without touching the files you have in your repo.

If you do `git status` now, it should show you all the changed files from this branch. Just add and commit them as usual with the new commit message you want for this branch. Optionally verify that the log looks like you expect with `git log`. When you're done, you can push your changes with `git push --force`.

File: storm-server/src/test/java/org/apache/storm/LocalStateTest.java [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+            LocalState ls1 = new LocalState(dir1_tmp.getPath());
+            LocalState ls2 = new LocalState(dir2_tmp.getPath());
+            Assert.assertTrue(ls1.snapshot().isEmpty());
+
+            ls1.put("a", globalStreamId_a);
+            ls1.put("b", globalStreamId_b);
+            Assert.assertTrue(ls1.snapshot().containsKey("a"));
+            Assert.assertTrue(ls1.snapshot().containsValue(globalStreamId_a));
+            Assert.assertTrue(ls1.snapshot().containsKey("b"));
+            Assert.assertTrue(ls1.snapshot().containsValue(globalStreamId_b));
```
Comment (revans2) [inline storm-server/src/test/java/org/apache/storm/LocalStateTest.java]: I would prefer to have to code more closely follow the test.

```
//(is (= {"a" gs-a "b" gs-b} (.snapshot ls1)))
Map<String, TBase> expected = new HashMap<>();
expected.put("a", globalStreamId_a);
expected.put("b", globalStreamId_b);
Assert.assertEquals(expected, ls1.snapshot());
```

File: storm-server/src/test/java/org/apache/storm/LocalStateTest.java [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+            Assert.assertEquals(globalStreamId_a, ls1.get("a"));
+            Assert.assertEquals(null, ls1.get("c"));
+            Assert.assertEquals(ls1.snapshot(), new LocalState(dir1_tmp.getPath()).snapshot());
```
Comment (revans2) [inline storm-server/src/test/java/org/apache/storm/LocalStateTest.java]: Similar here I would prefer to see it match the test
```
//(is (= {"a" gs-a "b" gs-b} (.snapshot (LocalState. dir1))))
Assert.assertEquals(expected, new LocalState(dir1_tmp.getPath()).snapshot());
```

**Comment by HeartSaVioR** [comment]
@kamleshbhatt 
Could you address @revans2 review comments? And if you have hard time to squash, please follow @srdo 's suggestion.

Or if it's also hard to follow, you can just follow my suggestion:

- create a new branch against current master
- make a change (since your changeset is easy to apply manually, you can do it without cherry-pick)
- commit (write down commit sha)
- get back to PR's branch
- `git reset --hard <commit sha>` 
- `git push --force`

**Comment by kamleshbhatt** [comment]
My Apology for the delay. Applied the review comments and created a new pull request
https://github.com/apache/storm/pull/2154/files.

Please take a look.

**Comment by srdo** [comment]
@kamleshbhatt I think it is easier to continue the review if you just force push your new commit onto this branch instead of creating a new one. GitHub will automatically update this PR when you do that.

**Comment by kamleshbhatt** [comment]
@srdo I was unable to squash the commits so created a new PR and applied all required changes.

**Comment by srdo** [comment]
If you do the following, your new branch gets copied onto this one.
`git checkout STORM-1290`
`git reset --hard STORM-1290_1`
`git push --force`

This sets the STORM-1290 branch to contain exactly the same commits as STORM-1290_1. When you then force push, the remote copy gets overwritten and this PR gets updated.

**Comment by kamleshbhatt** [comment]
@srdo Thanks for providing above steps. I applied these commands. It updated the  branch STORM-1290 but doesn't allow me to reopen this PR.

**Comment by srdo** [comment]
Weird. There should be a button marked "Reopen pull request" next to the comment box.

**Comment by srdo** [comment]
Okay, nevermind. The option is blocked if there's a force push while the PR is closed. https://github.com/isaacs/github/issues/361.

Sorry about that. We'll continue the review on the new PR then.
